libs/ebpfnetsh/programs.cpp

```diff
@@ -115,20 +141,35 @@ handle_ebpf_add_program(
         return ERROR_SUPPRESS_OUTPUT;
     }
 
-    // TODO(#188): add option to pin all programs in the object.
     struct bpf_program* program = bpf_program__next(nullptr, object);
     struct bpf_link* link;
-    result = ebpf_program_attach(program, (tags[1].bPresent ? &attach_type : nullptr), nullptr, 0, &link);
+    result = ebpf_program_attach(program, (tags[TYPE_INDEX].bPresent ? &attach_type : nullptr), nullptr, 0, &link);
     if (result != EBPF_SUCCESS) {
         std::cerr << "error " << result << ": could not attach program" << std::endl;
         return ERROR_SUPPRESS_OUTPUT;
     }
 
-    if (!pinned.empty()) {
-        if (bpf_program__pin(program, pinned.c_str()) < 0) {
-            std::cerr << "error " << errno << ": could not pin program" << std::endl;
+    if (pinned_type == PT_FIRST) {
+        if (pinpath.empty()) {
+            pinpath = bpf_program__name(program);
+        }
+        if (bpf_program__pin(program, pinpath.c_str()) < 0) {
+            std::cerr << "error " << errno << ": could not pin to " << pinpath << std::endl;
             return ERROR_SUPPRESS_OUTPUT;
         }
+    } else if (pinned_type == PT_ALL) {
+        // The pinpath specified is like a "directory" under which to pin programs.
+        if (!pinpath.empty()) {
+            pinpath += "/";
```

Comment:
should we not do the same i.e. append the program name to the provided path when pinned_type is "PT_FIRST"

Reply:
As far as I know, that's not what bpftool does.  My intent was to match bpftool behavior.

Reply:
https://www.mankier.com/8/bpftool-prog says:

> **bpftool prog load** pins only the first program from the _OBJ_ as _PATH_. **bpftool prog loadall** pins all programs from the _OBJ_ under _PATH_ directory.